File: BonafitApi/app/booking.py
from datetime import UTC, datetime, timedelta
from zoneinfo import ZoneInfo

from app.config import settings

MADRID = ZoneInfo(settings.timezone)

OCCUPIES_SLOT = {"pending", "confirmed"}
CONSUMES_SESSION = {"confirmed", "completed"}
POOL_CATEGORIES = {"entrenamiento-personal", "hipopresivos"}


def to_madrid(value: datetime) -> datetime:
    if value.tzinfo is None:
        value = value.replace(tzinfo=UTC)
    return value.astimezone(MADRID)


def iso_weekday(value: datetime) -> int:
    return to_madrid(value).isoweekday()


def parse_time_parts(value: str) -> tuple[int, int]:
    hours_s, minutes_s = value.split(":")
    return int(hours_s), int(minutes_s)


def at_local_time(day: datetime, hhmm: str) -> datetime:
    local_day = to_madrid(day).replace(hour=0, minute=0, second=0, microsecond=0)
    hours, minutes = parse_time_parts(hhmm)
    return local_day.replace(hour=hours, minute=minutes)


def add_minutes(value: datetime, minutes: int) -> datetime:
    return value + timedelta(minutes=minutes)


def start_of_local_day(value: datetime) -> datetime:
    return to_madrid(value).replace(hour=0, minute=0, second=0, microsecond=0)


def earliest_bookable_local_date(now: datetime, cutoff_time: str) -> datetime:
    cutoff = at_local_time(now, cutoff_time)
    day = start_of_local_day(now)
    if to_madrid(now) < cutoff:
        day = day + timedelta(days=1)
    else:
        day = day + timedelta(days=2)
    return day


def ranges_overlap(a_start: datetime, a_end: datetime, b_start: datetime, b_end: datetime) -> bool:
    return a_start < b_end and a_end > b_start


def occupies_trainer_slot(status: str) -> bool:
    return status in OCCUPIES_SLOT


def is_active_client_appointment(status: str) -> bool:
    return status in OCCUPIES_SLOT


def consumes_session(status: str) -> bool:
    return status in CONSUMES_SESSION


def session_delta(previous_status: str | None, next_status: str) -> int:
    before = 1 if previous_status and consumes_session(previous_status) else 0
    after = 1 if consumes_session(next_status) else 0
    return before - after


def is_bono_expired(expires_at: datetime | None, now: datetime) -> bool:
    if expires_at is None:
        return False
    exp = expires_at if expires_at.tzinfo else expires_at.replace(tzinfo=UTC)
    current = now if now.tzinfo else now.replace(tzinfo=UTC)
    return exp <= current


def is_bono_usable(remaining: int, expires_at: datetime | None, now: datetime) -> bool:
    return remaining > 0 and not is_bono_expired(expires_at, now)


def pick_preferred_bono(rows: list, now: datetime):
    usable = [row for row in rows if is_bono_usable(row.remaining_sessions, row.expires_at, now)]
    if not usable:
        return None

    def sort_key(row):
        exp = row.expires_at.timestamp() if row.expires_at else float("inf")
        return (exp, row.remaining_sessions)

    return sorted(usable, key=sort_key)[0]


def status_on_create(actor: str, instant_confirm: bool) -> str:
    if actor == "admin" or instant_confirm:
        return "confirmed"
    return "pending"


def status_on_client_reschedule(instant_confirm: bool) -> str:
    return "confirmed" if instant_confirm else "pending"


def list_availability_slots(
    *,
    duration_minutes: int,
    schedules: list,
    appointments: list,
    cutoff_time: str,
    now: datetime,
    from_dt: datetime,
    to_dt: datetime,
    actor: str,
    trainer_id: str | None = None,
    ignore_appointment_id: str | None = None,
) -> list[dict]:
    if duration_minutes <= 0:
        return []
    earliest = earliest_bookable_local_date(now, cutoff_time) if actor == "client" else None
    busy = [
        row
        for row in appointments
        if occupies_trainer_slot(row.status) and row.id != ignore_appointment_id
    ]
    days: list[datetime] = []
    cursor = start_of_local_day(from_dt)
    last = start_of_local_day(to_dt)
    while cursor <= last:
        days.append(cursor)
        cursor = cursor + timedelta(days=1)

    slots: list[dict] = []
    for day in days:
        if earliest and day < earliest:
            continue
        weekday = day.isoweekday()
        for schedule in schedules:
            if schedule.weekday != weekday:
                continue
            if trainer_id and schedule.trainer_id != trainer_id:
                continue
            window_start = at_local_time(day, schedule.start_time)
            window_end = at_local_time(day, schedule.end_time)
            slot_start = window_start
            while add_minutes(slot_start, duration_minutes) <= window_end:
                slot_end = add_minutes(slot_start, duration_minutes)
                taken = any(
                    row.trainer_id == schedule.trainer_id
                    and ranges_overlap(slot_start, slot_end, to_madrid(row.starts_at), to_madrid(row.ends_at))
                    for row in busy
                )
                if not taken:
                    slots.append(
                        {
                            "trainerId": schedule.trainer_id,
                            "startsAt": to_utc_iso(slot_start),
                            "endsAt": to_utc_iso(slot_end),
                        }
                    )
                slot_start = slot_end
    return slots


def slot_matches(slots: list[dict], trainer_id: str, starts_at: datetime, ends_at: datetime) -> bool:
    start_iso = to_utc_iso(starts_at)
    end_iso = to_utc_iso(ends_at)
    return any(
        slot["trainerId"] == trainer_id and slot["startsAt"] == start_iso and slot["endsAt"] == end_iso
        for slot in slots
    )


def to_utc_iso(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=UTC)
    utc = value.astimezone(UTC)
    millis = utc.microsecond // 1000
    return utc.strftime("%Y-%m-%dT%H:%M:%S") + f".{millis:03d}Z"


def parse_iso(value: str) -> datetime:
    normalized = value.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(normalized)
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed

File: BonafitDesk/src/app/core/booking.ts
import { AppointmentDto, AppointmentStatus, AvailabilitySlotDto } from '../models/appointment.dto';
import { BookingSettingsDto } from '../models/booking-settings.dto';
import { ClientBonoDto } from '../models/client-bono.dto';
import { ServiceDto } from '../models/service.dto';
import { TrainerDto } from '../models/trainer.dto';
import { TrainerScheduleDto } from '../models/trainer-schedule.dto';

export type BookingActor = 'client' | 'trainer';

export function occupiesTrainerSlot(status: AppointmentStatus): boolean {
  return status === 'pending' || status === 'confirmed';
}

export function isActiveClientAppointment(status: AppointmentStatus): boolean {
  return status === 'pending' || status === 'confirmed';
}

export function hasActiveClientAppointmentForService(
  rows: Pick<AppointmentDto, 'id' | 'serviceId' | 'status'>[],
  serviceId: string,
  ignoreAppointmentId?: string,
): boolean {
  return rows.some(
    (row) =>
      row.serviceId === serviceId &&
      row.id !== ignoreAppointmentId &&
      isActiveClientAppointment(row.status),
  );
}

export function pickNextClientAppointment<
  T extends Pick<AppointmentDto, 'status' | 'startsAt' | 'endsAt'>,
>(rows: T[], now: Date): T | undefined {
  const active = rows.filter((row) => isActiveClientAppointment(row.status));
  if (active.length === 0) {
    return undefined;
  }
  const nowMs = now.getTime();
  const notFinished = active.filter((row) => new Date(row.endsAt).getTime() >= nowMs);
  const pool = notFinished.length > 0 ? notFinished : active;
  const direction = notFinished.length > 0 ? 1 : -1;
  return [...pool].sort(
    (left, right) =>
      direction * (new Date(left.startsAt).getTime() - new Date(right.startsAt).getTime()),
  )[0];
}

export function isBonoExpired(expiresAt: string | null, now: Date): boolean {
  if (!expiresAt) {
    return false;
  }
  return new Date(expiresAt).getTime() <= now.getTime();
}

export function isBonoUsable(row: ClientBonoDto, now: Date): boolean {
  return row.remainingSessions > 0 && !isBonoExpired(row.expiresAt, now);
}

export function isGiftCredit(row: ClientBonoDto): boolean {
  return row.isGift === true;
}

export function pickPreferredBono(rows: ClientBonoDto[], now: Date): ClientBonoDto | null {
  const usable = rows.filter((row) => isBonoUsable(row, now));
  if (usable.length === 0) {
    return null;
  }
  return [...usable].sort((left, right) => {
    const expLeft = left.expiresAt ? new Date(left.expiresAt).getTime() : Number.POSITIVE_INFINITY;
    const expRight = right.expiresAt ? new Date(right.expiresAt).getTime() : Number.POSITIVE_INFINITY;
    if (expLeft !== expRight) {
      return expLeft - expRight;
    }
    return left.remainingSessions - right.remainingSessions;
  })[0];
}

export function isoWeekday(date: Date): number {
  const day = date.getDay();
  return day === 0 ? 7 : day;
}

export function parseTimeParts(value: string): { hours: number; minutes: number } | null {
  const match = /^(\d{2}):(\d{2})$/.exec(value);
  if (!match) {
    return null;
  }
  const hours = Number(match[1]);
  const minutes = Number(match[2]);
  if (hours > 23 || minutes > 59) {
    return null;
  }
  return { hours, minutes };
}

export function atLocalTime(day: Date, hhmm: string): Date {
  const parts = parseTimeParts(hhmm);
  const next = new Date(day);
  next.setHours(parts?.hours ?? 0, parts?.minutes ?? 0, 0, 0);
  return next;
}

export function addMinutes(date: Date, minutes: number): Date {
  return new Date(date.getTime() + minutes * 60_000);
}

export function startOfLocalDay(date: Date): Date {
  const next = new Date(date);
  next.setHours(0, 0, 0, 0);
  return next;
}

export function earliestBookableLocalDate(now: Date, cutoffTime: string): Date {
  const cutoff = atLocalTime(now, cutoffTime);
  const day = startOfLocalDay(now);
  if (now.getTime() < cutoff.getTime()) {
    day.setDate(day.getDate() + 1);
  } else {
    day.setDate(day.getDate() + 2);
  }
  return day;
}

export function isClientStartAllowed(startsAt: Date, now: Date, cutoffTime: string): boolean {
  return startOfLocalDay(startsAt).getTime() >= earliestBookableLocalDate(now, cutoffTime).getTime();
}

export function rangesOverlap(aStart: Date, aEnd: Date, bStart: Date, bEnd: Date): boolean {
  return aStart < bEnd && aEnd > bStart;
}

export function concurrentCapacityOf(trainerId: string, trainers?: Pick<TrainerDto, 'id' | 'concurrentCapacity'>[]): number {
  const found = trainers?.find((row) => row.id === trainerId);
  const value = found?.concurrentCapacity ?? 1;
  return value >= 1 ? value : 1;
}

export function overlappingOccupancy(
  busy: Pick<AppointmentDto, 'trainerId' | 'status' | 'startsAt' | 'endsAt'>[],
  trainerId: string,
  slotStart: Date,
  slotEnd: Date,
): number {
  return busy.filter(
    (row) =>
      row.trainerId === trainerId &&
      occupiesTrainerSlot(row.status) &&
      rangesOverlap(slotStart, slotEnd, new Date(row.startsAt), new Date(row.endsAt)),
  ).length;
}

export function slotTakenForTrainer(
  busy: Pick<AppointmentDto, 'trainerId' | 'status' | 'startsAt' | 'endsAt'>[],
  trainerId: string,
  slotStart: Date,
  slotEnd: Date,
  concurrentCapacity: number,
): boolean {
  const capacity = concurrentCapacity >= 1 ? concurrentCapacity : 1;
  return overlappingOccupancy(busy, trainerId, slotStart, slotEnd) >= capacity;
}

export function slotTakenForClient(
  busy: Pick<AppointmentDto, 'clientId' | 'status' | 'startsAt' | 'endsAt'>[],
  clientId: string,
  slotStart: Date,
  slotEnd: Date,
): boolean {
  return busy.some(
    (row) =>
      row.clientId === clientId &&
      occupiesTrainerSlot(row.status) &&
      rangesOverlap(slotStart, slotEnd, new Date(row.startsAt), new Date(row.endsAt)),
  );
}

export function statusOnCreate(actor: BookingActor, instantConfirm: boolean): AppointmentStatus {
  if (actor === 'trainer' || instantConfirm) {
    return 'confirmed';
  }
  return 'pending';
}

export function statusOnClientReschedule(instantConfirm: boolean): AppointmentStatus {
  return instantConfirm ? 'confirmed' : 'pending';
}

export interface AvailabilityInput {
  service: ServiceDto;
  schedules: TrainerScheduleDto[];
  appointments: AppointmentDto[];
  settings: BookingSettingsDto;
  now: Date;
  from: Date;
  to: Date;
  actor: BookingActor;
  trainerId?: string;
  ignoreAppointmentId?: string;
  trainers?: Pick<TrainerDto, 'id' | 'concurrentCapacity'>[];
  clientId?: string;
}

export function listAvailabilitySlots(input: AvailabilityInput): AvailabilitySlotDto[] {
  const duration = input.service.durationMinutes;
  if (duration <= 0) {
    return [];
  }
  const earliest =
    input.actor === 'client' ? earliestBookableLocalDate(input.now, input.settings.nextDayCutoffTime) : null;
  const busy = input.appointments.filter(
    (row) => occupiesTrainerSlot(row.status) && row.id !== input.ignoreAppointmentId,
  );
  const days: Date[] = [];
  const cursor = startOfLocalDay(input.from);
  const last = startOfLocalDay(input.to);
  while (cursor.getTime() <= last.getTime()) {
    days.push(new Date(cursor));
    cursor.setDate(cursor.getDate() + 1);
  }
  const slots: AvailabilitySlotDto[] = [];
  for (const day of days) {
    if (earliest && day.getTime() < earliest.getTime()) {
      continue;
    }
    const weekday = isoWeekday(day);
    for (const schedule of input.schedules) {
      if (schedule.weekday !== weekday) {
        continue;
      }
      if (input.trainerId && schedule.trainerId !== input.trainerId) {
        continue;
      }
      const windowStart = atLocalTime(day, schedule.startTime);
      const windowEnd = atLocalTime(day, schedule.endTime);
      let slotStart = new Date(windowStart);
      while (addMinutes(slotStart, duration).getTime() <= windowEnd.getTime()) {
        const slotEnd = addMinutes(slotStart, duration);
        const taken =
          slotTakenForTrainer(
            busy,
            schedule.trainerId,
            slotStart,
            slotEnd,
            concurrentCapacityOf(schedule.trainerId, input.trainers),
          ) ||
          Boolean(
            input.clientId && slotTakenForClient(busy, input.clientId, slotStart, slotEnd),
          );
        if (!taken) {
          slots.push({
            trainerId: schedule.trainerId,
            startsAt: slotStart.toISOString(),
            endsAt: slotEnd.toISOString(),
          });
        }
        slotStart = slotEnd;
      }
    }
  }
  return slots;
}

export function slotMatches(
  slots: AvailabilitySlotDto[],
  trainerId: string,
  startsAt: string,
  endsAt: string,
): boolean {
  return slots.some(
    (slot) => slot.trainerId === trainerId && slot.startsAt === startsAt && slot.endsAt === endsAt,
  );
}

export function remainingSessionsDelta(
  fromStatus: AppointmentStatus | null,
  toStatus: AppointmentStatus,
): number {
  const before = fromStatus && consumesSession(fromStatus) ? 1 : 0;
  const after = consumesSession(toStatus) ? 1 : 0;
  return before - after;
}

function consumesSession(status: AppointmentStatus): boolean {
  return status === 'confirmed' || status === 'completed';
}

export function canCancelAppointment(
  status: AppointmentStatus,
  startsAt: Date,
  now: Date,
  cutoffTime: string,
): boolean {
  if (!isActiveClientAppointment(status)) {
    return false;
  }
  return isClientStartAllowed(startsAt, now, cutoffTime);
}

export function canClientConfirmAppointment(status: AppointmentStatus): boolean {
  return status === 'pending';
}

export function canAdminCancelAppointment(status: AppointmentStatus): boolean {
  return status === 'pending' || status === 'confirmed';
}
